### PR TITLE
feat(conformance): 175c0a harness honesty + debt-ratchet foundation (#175)

### PR DESCRIPTION
**Pull request by kortiene**
> → open PR) requires the maintainer's explicit word, as does merging.

**Base:** main @ `4eda978` · **Branch:** `feat/wave2-175c0a-harness-honesty`
· **Commits:** `5899d6a` (foundation) → `e99bb89` (tests/CI/README) →
`9cee5cb` (probe-caught: ratchet mismatches were swallowed) → `4b7ef69`
(self-caught: --print-ledger 64 KiB pipe truncation) → `af936f9`
(review round: deterministic oversize-frame close; dead captures; stale
README sentence).

## What

The protocol-v2 conformance harness's honesty foundation, per the accepted
175c0a contract (wave-2 board, 2026-08-23). Six changes, no protocol/core
semantics, no new selector, no scenario, no topology:

1. **Metadata honesty** — `conformance/v2/harness/daemon.mjs` binds
   `pf.storage_generation` from the portfile (it read `pf.protocol`; both
   axes are `2` today, so every generation-gated replay was green by
   coincidence). Proven by a stub-daemon unit test with deliberately
   diverged axes AND a live probe with a rebuilt daemon
   (`STORAGE_GENERATION=3`): the fixed harness passes, the old binding
   fails with the 426 `storage_generation_mismatch` refusal.
2. **Selector hygiene** — `main.mjs` exits 2 for an unknown `--case`
   (unknown names listed), a duplicate `--case`, `--case` together with
   `--filter`, and a zero-match `--file`/`--filter` — every guard fires
   before the Runner (and any daemon) is constructed. The checker's
   CI-selector existence/unique guard is untouched.
3. **Requires honesty** — a well-formed `requires` token the runner does
   not establish (`link:*`, `room:removed|foreign`, `member:agent|
   non_agent`, `daemon:restartable`, `observe:*`, `control:concurrency`,
   `resource:large_file|fetched_file`, `fault:*`) now fails the case
   BEFORE staging, naming the token — never a silent half-run (a full
   corpus replay previously half-ran ~180 such cases).
4. **One capability classification** —
   `conformance/v2/harness/capabilities.mjs` is the single
   executable/documented-only list for control verbs plus the runner's
   implemented-requires set. The checker refuses to validate on drift; the
   README control table carries the classification as a column and a unit
   test compares row by row; the runner's refusal names the
   classification. Shape errors fail for both classes. No control is newly
   implemented.
5. **Global structural ratchet** — the assert-shape, empty-assert, and
   variable-binding rules (previously files.json-only) now apply to every
   domain, plus new shape rules for the padded/bare-envelope `send` forms.
   Debt in the unselected corpus is grandfathered by an EXACT multiset
   ledger (`scripts/v2-corpus-debt-ledger.json`, derived from the parsed
   corpus: **580 structural fingerprints across 161 unselected cases**)
   keyed by rule-id + semantic location (`file|case|step N[.assert[i]]`):
   a new or moved fingerprint fails, a removal forces the ledger to
   shrink, and a missing ledger means strict mode. The CI-selected slice
   must stay debt-free.
6. **Note ratchets, parsed by structural attachment** (never raw grep):
   after this change UNMAPPED 241 / unrepaired 32 / split 55; dual notes
   count in both facets and aggregate summaries use the disjoint
   partition.

Also repairs the 26 CI-selected cases to zero debt (9 UNMAPPED + 3
unrepaired notes and all 10 structural findings: 5 empty asserts, 2
array-valued `ne`, 1 malformed `$max_frame + 1` expression, 2 unbound
save-sources) — see "Selected-case repairs" below.

## Two false-greens this found (and fixed)

- **The max-frame boundary cases executed nothing.** The `send` verb's
  padded-envelope form (`{envelope, pad_field, pad_byte,
  pad_to_total_frame_bytes}`) was implemented nowhere: the runner
  serialized the instruction object itself (no top-level id, so the daemon
  discarded it), and the fixtures' `frame.hello.limits.*` save paths — a
  v1 remnant — never resolved, silently capturing `undefined`. This PR
  implements `buildPaddedEnvelope` (grows the pad field until the whole
  serialized frame is exactly the target, `$max_frame` or
  `{"$add": ["$max_frame", 1]}`), unwraps bare-envelope sends, and fixes
  the save paths. The at-limit case now sends a real 128 MiB frame, awaits
  its correlated reply, and asserts the connection stays open; the
  one-past case observes close code 4005. A raw-send write aborted by the
  daemon's close is recorded rather than thrown, and a case-end guard
  refuses a green verdict unless a later step observed the connection
  state.
- **The ratchet's own first draft swallowed its verdicts** — caught by
  negative probe: the ledger-comparison mismatches were routed through a
  rule-bearing `fail()` that records ratcheted findings instead of
  problems, so an injected unbound variable left the checker green. Fixed;
  the probe (unbound `$var` in rooms.json) now fails with the exact
  fingerprint named.

## Selected-case repairs (conformance/v2/subject-daemon.json, 10 cases)

- empty asserts → real assertions: `frame.err` absent (the dialect's
  `must_not_error_with`/`assert_over_session` are subsumed by "no error
  arm at all"; the `err` root itself always resolves, so `frame.err` is
  the expressible form), `$n` type/gte (outstanding: the served bound is
  a live positive integer; the at/past burst legs need the
  documented-only `start_transfers` control and a static burst would
  hardcode the served constant against the independence rule — noted in
  the fixture).
- `forbid_keys_deep` → `exact_keys` where the record pins the reply's
  shape (subject.ensure) and 9 per-key `absent` where it deliberately does
  not (daemon.stop) — the deep form has no closed-op equivalent and is
  documented as such in the fixture note.
- `distinct_values` → `out.subject_id ne $did`; array-valued `ne` split
  into one `ne` per forbidden value; `$out`/`$err` save roots corrected;
  two transport-closed dialect steps folded into the `process_exited`
  proofs (the observe vocabulary cannot say "closed with any code" for a
  named session without pinning a number the record never names).
- 4 split-only provenance notes retired with their folded steps; the 2
  that remain sit on executable steps, as the contract allows.

## Gate evidence (worktree, base 4eda978)

| Gate | Result |
|---|---|
| `node scripts/check-v2-corpus.mjs` | OK — 348 cases / 2150 steps (two folded steps; no lockstep file states a step count), 580 fingerprints grandfathered, notes 241/32/55, ledger exact |
| `node --test scripts/check-v2-corpus.test.mjs` | 19/19 |
| `node --test conformance/v2/harness/stream.test.mjs` | 47/47 |
| `node --test conformance/v2/harness/harness.test.mjs` | 18/18 (new) |
| Live replay, the CI 26 (`node main.mjs target/debug/jeliyad --case …`) | **26/26 PASS** (now including the two real max-frame cases, ~2×128 MiB loopback) |
| `cargo test --locked --workspace` | green |
| `cargo clippy --locked --workspace -- -D warnings` | green (no Rust source changed on the branch) |
| `cargo fmt --all --check` | green |
| `node scripts/check-docs.mjs` | OK |

## Negative probes (each red for the intended reason; reverts
context-unique; full suite re-run green after the round)

| Probe | Intended reason (observed) |
|---|---|
| protocol ≠ storage_generation (rebuilt daemon, old binding) | 426 storage_generation_mismatch refusal; fixed binding passes |
| typo `--case` | exit 2, unknown name listed, before any daemon |
| duplicate `--case` | exit 2, duplicate named |
| `--case` + `--filter` | exit 2, mutual exclusion |
| zero-match `--file` / `--filter` | exit 2, available files / case count |
| valid-but-unimplemented require (`link:up` case) | pre-stage refusal naming the token (corpus-wide: ~180 loud errors) |
| malformed require (`link:upstairs`) | "namespace link permits only [up, down, relay, slow]" |
| unbound `$var` outside files.json | "new/moved structural debt (bind:never-bound) … not in the ledger" |
| empty assert outside files.json | "new/moved structural debt (empty-assert) …" |
| equal-count fingerprint swap | BOTH the new fingerprint and the "shrink the ledger" demand fire; count stays 580 |
| new UNMAPPED marker (unselected case) | "new/moved unmapped note marker …" |
| selected case carrying debt | "CI-selected case carries … debt-free" |
| classification drift (checker keys / README class / runner live) | checker drift fail / 7 harness tests red / live ERROR naming the classification |

## Adversarial review (claim-attack, read-only)

Two disjoint reviewer briefs. Findings and dispositions:
- **FIXED** — false-red flake: the one-past max-frame case lost a TCP race
  (~11/12 runs): the daemon refuses an oversize frame from its header while
  a one-shot 128 MiB write is in flight; the daemon-side unread backlog
  provokes an RST that discards the 4005 close frame before the client
  parses it (reads 1006). `sendPaddedTrickled` now writes the frame HEADER
  first and pauses: a refusal arrives with zero unread backlog
  (deterministic); only an accepted header streams the payload. The fix
  also caught a missing mask copy in the hand-rolled header (now a unit
  test). Post-fix stability: at-limit 4/4, one-past 4/4 + 3/3, live 26/26.
- **FIXED** — dead captures (`first_out`, `$e_wellformed`) removed: they
  bound values no later step reads.
- **FIXED** — stale README executor-row sentence (claimed transport_drop
  unimplemented, contradicting its own cell and CI).
- **DEFERRED (reason recorded)** — the `outstanding_requests…at_and_past`
  case name overclaims (its burst legs need the documented-only
  `start_transfers` control); the fixture note discloses this; renaming
  would churn the CI selector list, which this slice's scope forbids.
- Reviewer verdicts on the honesty claims: ledger counts independently
  reproduced (580/161; 241/32/53); facet-vs-disjoint arithmetic reconciles
  exactly (328 instances − 19 dual = 309 locations); no independence
  violation found in any repaired pin; the `frame.err`-absent and
  `exact_keys` derivations verified against the assert engine and the
  record.

## Reviewer notes

- The ledger (`scripts/v2-corpus-debt-ledger.json`) is DERIVED
  (`--print-ledger`) and then deliberately written; reviewers should spot
  check that its counts (580/161; facets 241/32/55) match the checker's
  summary line.
- The contract's requires-throw parenthetical listed `daemon:fresh` among
  throw candidates; the binding record's wording ("valid-but-unimplemented")
  governs — `daemon:fresh` is established (fresh daemon per case), and
  throwing on it would contradict the 26/26 gate. `control:limits`/
  `control:clock` likewise (served limits surfaced / capped real wait);
  the override verb `set_limit` itself stays documented-only.
- The `outstanding_requests` case remains honestly partial: its at/past
  burst legs need `start_transfers {count}` (documented-only here); the
  fixture now says so instead of carrying a vacuous assert.

## Remaining scope

This is slice 175c0a of #175's remaining scope (175b1/2, 175c0b,
175c1a/1b, 175c2–175c6, 175d). **merging this does not complete issue
#175** — the honesty/debt-ratchet foundation lands here; the RealNetwork
topology, scenario slices, and Android CI follow in their own PRs.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -458,6 +458,17 @@ jobs:
...
+      # 175c0a: the harness's own honesty foundations run as unit tests in
+      # CI — selector hygiene, requires honesty, the shared capability
+      # classification (drift vs the README control table and the checker's
+      # vocabulary), the portfile storage_generation binding, and the
+      # padded-frame builder — alongside the stream codec suite. Before this,
+      # only shape validation and the live replay were wired.
+      - name: Protocol-v2 conformance harness unit tests
+        run: |
+          set -euo pipefail
+          node --test conformance/v2/harness/stream.test.mjs conformance/v2/harness/harness.test.mjs
```
[the remark this answers is not in the document]
Comment (kortiene) [reply .github/workflows/ci.yml]: Already fixed in `4cf6578` (pushed before this comment landed — it references the pre-fix revision): the step now runs `npm ci --ignore-scripts` inside `conformance/v2/harness` before the tests, mirroring the replay step. Verified on CI: the second run's "Protocol-v2 conformance harness unit tests" step concluded `success` (steps API), 66/66 tests.

File: conformance/v2/harness/capabilities.mjs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  'daemon:self',
+  'daemon:fresh',
+  'daemon:second',
+  'room:plain',
+  'room:live',
+  'room:quiescent',
+  'room:left',
+  'room:with_history',
+  'member:b',
+  'member:c',
```
[the remark this answers is not in the document]
Comment (kortiene) [reply conformance/v2/harness/capabilities.mjs]: Confirmed on all counts and fixed in `8d38f11`. Verified against `#establishRequires`: `room:quiescent`/`room:with_history` stage identically to `room:live` (activate; no history authored), `room:left` creates `` with no join/leave, and `subject:member_b`/`c` don't match the routing regex so they land on the primary daemon whose single subject is the authority's. All five tokens removed from `REQUIRES_IMPLEMENTED` — the guard now refuses those cases loudly pre-stage (probed: a `room:with_history` case and a `room:left` case both ERROR naming the token). The CI-selected 26 use none of these tokens (verified programmatically), so the live slice is unaffected: 26/26 green after the change. The unfaithful staging itself is pre-existing (the board review recorded it as hazards #3/#4); making its tokens honestly refuse is exactly this slice's charter — real repair of the staging is 175c0b/175c2 territory.

File: scripts/check-v2-corpus.mjs [hunk 3]
```diff
@@ -1598,10 +1760,191 @@ if (isObject(manifest)) {
...
+  set.add(f.facet);
+  disjoint.set(key, set);
+}
+const disjointCounts = {};
+for (const facets of disjoint.values()) {
+  const tag = [...facets].sort().join("+");
+  disjointCounts[tag] = (disjointCounts[tag] ?? 0) + 1;
+}
+const facetCounts = {};
+for (const f of noteFindings) facetCounts[f.facet] = (facetCounts[f.facet] ?? 0) + 1;
```
[the remark this answers is not in the document]
Comment (kortiene) [reply scripts/check-v2-corpus.mjs]: Confirmed and fixed in `8d38f11`: both aggregates (and the human-readable summary line) now filter `selectedCases` before construction. Recomputed: `unselected_note_facet_counts` = {split: 53, unmapped: 241, unrepaired: 32}, matching the ledger exactly; the 2 selected markers still surface under the separate `selected_note_markers` field.

File: conformance/v2/harness/session.mjs [hunk 4]
```diff
@@ -38,6 +38,90 @@ export function requestId() {
...
+  cur[leaf] = '';
+  const base = Buffer.byteLength(serializeWithRaw(envelope), 'utf8');
+  const originalBytes = Buffer.byteLength(JSON.stringify(original), 'utf8') - 2; // minus quotes
+  const fill = target - base - originalBytes;
+  if (fill < 0) {
+    throw new AssertFailure(
+      `pad_to_total_frame_bytes ${target} is smaller than the envelope itself (${base + originalBytes} bytes)`,
+    );
+  }
+  cur[leaf] = original + pad_byte.repeat(fill);
```
[the remark this answers is not in the document]
Comment (kortiene) [reply conformance/v2/harness/session.mjs:149]: Confirmed and fixed in `8d38f11` — took the restriction option: `buildPaddedEnvelope` and the checker's `send-shape:pad-byte` rule now both require a pad byte that is exactly one UTF-8 byte AND unescaped by JSON (`Buffer.byteLength === 1 && JSON.stringify(pad) === quoted pad`), with the failure message explaining why. A unit test pins the rejections (`\"`, `\\\\`, `é`, an emoji, multi-char, empty).

File: conformance/v2/harness/session.mjs [hunk 5]
```diff
@@ -38,6 +38,90 @@ export function requestId() {
...
 
+/** Build one masked client text-frame header for a payload of `length`
+ * bytes: FIN+text, MASK set, 64-bit extended length, then the 4 mask bytes.
+ * Pure — unit-tested (the mask MUST land at offset 10; a zero mask with a
+ * masked payload is protocol garbage the daemon can only abort). */
+export function buildMaskedFrameHeader(length, mask) {
+  const header = Buffer.alloc(14);
+  header[0] = 0x81;
+  header[1] = 0x80 | 127;
+  header.writeBigUInt64BE(BigInt(length), 2);
```
[the remark this answers is not in the document]
Comment (kortiene) [reply conformance/v2/harness/session.mjs]: Confirmed and fixed in `8d38f11`: `buildMaskedFrameHeader` now emits the canonical form per range — 0–125 inline (`0x80|len`, mask at offset 2), 126–65535 as 16-bit (`mask at offset 4`), else the 64-bit form — and rejects negative/non-integer lengths. A unit test pins all three ranges plus the mask offsets. The daemon's tungstenite happened to accept the non-minimal form (which is why the boundary cases passed), but a stricter peer would have exercised the wrong refusal, as you note. Live slice re-verified 26/26 with the canonical encoding, and the one-past close-code case re-checked for stability (2/2).
